src/cli/rustup_mode.rs

```diff
@@ -1159,9 +1166,75 @@ fn component_add(cfg: &Cfg, m: &ArgMatches<'_>) -> Result<()> {
         let new_component = Component::new_with_target(component, false)
             .unwrap_or_else(|| Component::new(component.to_string(), target.clone(), true));
 
-        toolchain.add_component(new_component)?;
-    }
+        if m.is_present("allow-downgrade") {
+            let toolchain_date = toolchain
+                .get_manifest()
+                .ok()
+                .and_then(|m| m.map(|m| m.date));
+            let toolchain_date = toolchain_date.as_deref();
+            let downgrade_date = utc_from_manifest_date(toolchain_date.unwrap())
+                .unwrap_or_else(|| panic!("Malformed manifest date: {:?}", toolchain_date))
```

Comment:
then this should become
```suggestion
                .unwrap_or_else(|| panic!("Malformed manifest date: {}", toolchain_date))
```